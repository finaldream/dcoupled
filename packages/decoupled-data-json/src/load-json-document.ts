import fs from 'fs';
import util from 'util';
import { resolve } from 'path';

const readFile = (fileName: string): Promise<string> => util.promisify(fs.readFile)(fileName, 'utf8');

export const loadJsonDocument = async (
    documentName: string,
    docDir: string = 'content'
): Promise<AnyObject> => {

    const slug = documentName === '/'
        ? 'index'
        : documentName;

    let content;
    try {
        content = await readFile(resolve(docDir, `${slug}.json`));
    } catch (e) {
        return {
            error: 'not-found',
            message: e.message,
        };
    }


    let result;
    try {
        result = JSON.parse(content);
    } catch (e) {
        return {
            error: 'syntax-error',
            message: e.message,
        };
    }

    result.slug = slug;

    return result;

};
