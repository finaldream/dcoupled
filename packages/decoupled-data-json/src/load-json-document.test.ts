import { loadJsonDocument } from './load-json-document';

describe('loadJsonDocument()', () => {

    test('should resolve "/" to "index"', async () => {

        const content = await loadJsonDocument('/', 'fixtures');
        expect(content.title).toBe('Homepage');
        expect(content.slug).toBe('index');
        expect(content).toMatchSnapshot();

    });


    test('should parse complex Json', async () => {

        const layout = await loadJsonDocument('layout', 'fixtures');
        expect(layout.title).toBe('Layout Test');
        expect(layout.slug).toBe('layout');
        expect(layout).toMatchSnapshot();

    });

    test('should receive an error when document is broken', async () => {

        const content = await loadJsonDocument('broken', 'fixtures');
        expect(content.error).toBe('syntax-error');
        expect(content.message).toContain('Unexpected token');

    });


    test('should receive an error when document does not exist', async () => {

        const content = await loadJsonDocument('non-existent', 'fixtures');
        expect(content.error).toBe('not-found');
        expect(content.message).toContain('ENOENT:');

    });

});
