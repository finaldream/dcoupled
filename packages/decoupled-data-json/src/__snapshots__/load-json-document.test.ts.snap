// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadJsonDocument() should parse complex Json 1`] = `
Object {
  "date": "2019-08-07T18:30:00.000Z",
  "layout": Array [
    Object {
      "type": "text",
      "value": Array [
        "<h2>Hello World!</h2>",
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam ex odio, elementum non luctus eget, mollis vitae orci.",
        "Quisque dapibus purus eget turpis accumsan congue. Quisque blandit dui non purus hendrerit, vel ornare eros dignissim.",
        "Integer malesuada velit consequat dolor vulputate, eu bibendum lectus suscipit. Vivamus dapibus, ipsum et lobortis lacinia,",
        "ex elit tristique ex, quis placerat lorem arcu quis ligula. Aliquam orci tellus, placerat eget velit at, egestas lobortis eros.",
      ],
    },
    Object {
      "description": "Duis a viverra est. **Pellentesque dictum purus neque**, nec efficitur lectus malesuada nec. In sem sapien, venenatis nec erat eu, blandit scelerisque magna.",
      "title": "Services",
      "type": "services",
      "value": Object {
        "CI / CD": "/ci-cd",
        "Full-Stack Development": "/full-stack-development",
        "Microservices": "/microservices",
        "Mobile Development": "/mobile-development ",
        "Single Page Applications": "/single-page-applications",
        "Static Site Generators": "/static-site-generator",
        "UI/UX": "/ui-ux",
      },
    },
    Object {
      "title": "Workflow",
      "type": "workflow",
      "value": Array [
        "Analyse & Planning",
        "Development",
        "Maintenance",
      ],
    },
    Object {
      "title": "Technologies",
      "type": "tech",
      "value": Array [
        "Node.js",
        "Typescript",
        "React.js",
        "React Native",
        "Javascript",
        "Redux",
        "Jest",
        "GraphQL",
      ],
    },
    Object {
      "description": "Duis a viverra est. **Pellentesque dictum purus neque**, nec efficitur lectus malesuada nec. In sem sapien, venenatis nec erat eu, blandit scelerisque magna.",
      "title": "Contact",
      "type": "contact",
    },
  ],
  "slug": "layout",
  "title": "Layout Test",
}
`;

exports[`loadJsonDocument() should resolve "/" to "index" 1`] = `
Object {
  "content": Array [
    "<h2>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</h2>",
    "Aliquam ex odio, elementum non luctus eget, mollis vitae orci. Quisque dapibus purus eget turpis accumsan congue.",
    "Quisque blandit dui non purus hendrerit, vel ornare eros dignissim. Integer malesuada velit consequat dolor vulputate, eu bibendum lectus suscipit.",
    "Vivamus dapibus, ipsum et lobortis lacinia, ex elit tristique ex, quis placerat lorem arcu quis ligula. Aliquam orci tellus, placerat eget velit at,",
    "egestas lobortis eros. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Integer ut odio iaculis, vestibulum elit vel",
    "sagittis nisl. Nam convallis quam id massa convallis mollis. Phasellus ut nunc ullamcorper, placerat lacus non, viverra libero. Curabitur semper urna tortor.",
    "Phasellus varius aliquam odio eu malesuada. Nunc tempus egestas ante, non mollis purus blandit a. Etiam ac vulputate elit.",
  ],
  "date": "2019-10-01T10:33:18.754Z",
  "slug": "index",
  "title": "Homepage",
}
`;
